Introduce MockPrimaryNode for the domain testing framework

NingLin-P opened the pull request:
This PR is part of https://github.com/subspace/subspace/issues/1213.

This PR introduces `MockPrimaryNode` for the domain testing framework, each commit contains the changes of one functionality/API that `MockPrimaryNode` bring, please refer to the tests of `domain-client-executor` or `subspace-fraud-proof` in https://github.com/subspace/subspace/tree/domain-testing for the usage of these functionalities/APIs.

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

liuchengxu reviewed (commented):
It's pretty inconvenient to review with new codes completely in one PR but their usages are in another branch. Can you restructure this PR with a minimal new test primary node in the first commit, and then update the tests incrementally by adding new apis necessary for the tests, so that it's easier to see what problems the new apis are solving accordingly.

NingLin-P committed:
Introduce MockBlockImport to MockMockPrimaryNode

MockBlockImport is used to mock the block import notification interface of SubspaceBlockImport
It is mostly port from SubspaceBlockImport with all the consensus related logic removed.

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Mock block production in MockPrimaryNode

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Integrate MockPrimaryNode with domain-test-service

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Integrate MockPrimaryNode with test test_executor_full_node_catching_up

Signed-off-by: linning <linningde25@gmail.com>

liuchengxu commented on crates/sc-consensus-subspace/src/notification.rs:
```diff
@@ -29,7 +29,7 @@ type SharedNotificationSenders<T> = Arc<Mutex<Vec<TracingUnboundedSender<T>>>>;
 /// The sending half of the Subspace notification channel(s).
 #[derive(Clone)]
-pub(crate) struct SubspaceNotificationSender<T: Clone + Send + Sync + fmt::Debug + 'static> {
+pub struct SubspaceNotificationSender<T: Clone + Send + Sync + fmt::Debug + 'static> {
```
Hmm, I recall Nazar had a concern about making them public, which opens the door for some bad practices https://github.com/subspace/subspace/pull/926#discussion_r1026137365. Since `MockBlockImport` is from `SubspaceBlockImport` anyway, we may see whether `MockBlockImport` can be moved into `sc-consensus-subspace`. @nazar-pc thoughts?

NingLin-P replied:
Moving `MockBlockImport` into `sc-consensus-subspace` may not be enough since the slot notification also uses `SubspaceNotificationSender`. But I realized that the executor only required a stream produce item `(NumberFor<PBlock>, mpsc::Sender<()>)` and it doesn't have to be `SubspaceNotificationStream`, thus I use `TracingUnboundedReceiver` instead and remove the dependency of `sc-consensus-subspace`.

liuchengxu commented on test/subspace-test-service/src/mock.rs:
```diff
@@ -0,0 +1,397 @@
+    }
+
+    /// Import block
+    pub async fn import_block(
```
I'm not seeing this public API and `build_block` above used in the tests, are they used in your later commits that are not included in this PR? Better not make them public too early otherwise it's hard to tell what's exactly needed/used at this moment, we can expose them anytime later when needed. With proper APIs around `produce_block`, I even feel they don't have to be exposed eventually.

NingLin-P replied:
> are they used in your later commits that are not included in this PR?

Right, it is a workaround for injecting the `DigestItem::RuntimeEnvironmentUpdated` into the block header. I also thought of building an API based on `produce_block` but the inherent digest approach doesn't work for `DigestItem::RuntimeEnvironmentUpdated` somehow, anyway we can discuss it in later PRs.

liuchengxu reviewed (commented):
Make sense overall as a start, looking forward to more integrations of the new testing into the old tests. 

One thing I see is that no network for the primary chain which might be inconvenient for the messenger tests as the relayer then needs to send the message to the target domain node directly by hand, but I guess we can always add it later if necessary.

vedhavyas commented on test/subspace-test-service/src/mock.rs:
```diff
@@ -0,0 +1,224 @@
+
+// `MockBlockImport` is mostly port from `sc-consensus-subspace::SubspaceBlockImport` with all
+// the consensus related logic removed.
+struct MockBlockImport<Inner, Client, Block: BlockT> {
```
Can't we use the BlockChainvents trait implementated by client for this ?

NingLin-P replied:
We may be unable to do that, `MockBlockImport` is used to mock `sc-consensus-subspace::SubspaceBlockImport` which produces item `(NumberFor<PBlock>, mpsc::Sender<()>)`, it is mainly used for the throttling purpose.

vedhavyas replied:
ahh I see.

vedhavyas commented on domains/test/service/src/lib.rs:
```diff
@@ -615,3 +743,16 @@ pub async fn run_primary_chain_validator_node(
+    key: Sr25519Keyring,
+    base_path: BasePath,
+) -> subspace_test_service::mock::MockPrimaryNode {
+    subspace_test_service::mock::MockPrimaryNode::run_mock_primary_node(
```
seems wierd seeing the primary node start under domains :)
any reason why the wrapper here ?

NingLin-P replied:
Yeah, you are right, it is quite weird. And it seems to make more sense to use `MockPrimaryNode ::run_mock_primary_node` directly in the test.

> any reason why the wrapper here ?

It is just a mimic of the `run_primary_chain_validator_node`.

vedhavyas reviewed (commented):
Great work! Some minor nits.

I have a wild thought. Since we are aiming to test as close to the production node as possible. Can't we just run a mock farmer node and provide mock solutions. We can even mock verify solution. With Nazar's changes to add this verify_solution as Host function, we can create a mock host function instead.

NingLin-P committed:
Introduce MockPrimaryNode to the domain testing framework

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Mock slot production in MockPrimaryNode

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Merge branch 'main' into mock-primary-node

NingLin-P committed:
Merge branch 'main' into mock-primary-node
